Fix npm access error(#665)

DerTimonius opened the pull request:
<!--

Thanks for submitting a pull request 🙌

**Note:** Please don't create a pull request which has significant changes (i.e. adds new functionality or modifies existing one in a non-trivial way) without creating an issue first.

Try to limit the scope of your pull request and provide a general description of the changes. If this fixes an open issue, link to it in the following way: `Fixes #321`. New features and bug fixes should come with tests.

-->

<details>
<summary>Thank you for considering this PR</summary>
This is my first contribution to open source, so if this PR description is a bit too much, please let me know 😃 


</details>

This PR fixes the [``npm access`` error](https://github.com/sindresorhus/np/issues/665) that occurs after updating npm to version 9.0.0.

```bash
✖ Command failed with exit code 1: npm access ls-collaborators @upleveled/eslint-config-upleveled
npm ERR! code EUSAGE
npm ERR!
npm ERR! ls-collaborators is not a valid access command
npm ERR!
npm ERR! Set access level on published packages
npm ERR!
npm ERR! Usage:
npm ERR! npm access list packages [<user>|<scope>|<scope:team> [<package>]
npm ERR! npm access list collaborators [<package> [<user>]]
npm ERR! npm access get status [<package>]
npm ERR! npm access set status=public|private [<package>]
npm ERR! npm access set mfa=none|publish|automation [<package>]
npm ERR! npm access grant <read-only|read-write> <scope:team> [<package>]
npm ERR! npm access revoke <scope:team> [<package>]
npm ERR!
npm ERR! Options:
npm ERR! [--json] [--otp <otp>] [--registry <registry>]
npm ERR!
npm ERR! Run "npm help access" for more info
```
According to the [npm documentation](https://docs.npmjs.com/cli/v9/commands/npm-access) the ``npm access`` subcommands have been changed from ``ls-collaborators`` to ``list collaborators``.

Switching
```javascript
const args = ['access', 'ls-collaborators', packageName];
```
to
```javascript
const args = ['access', 'list collaborators', packageName];
```
did not work, it keeps failing with the same errors.
Changing the execa arguments (on line 77) from ``npm`` to ``npm access list collaborators`` while stripping the ``args`` variable eliminated the error.


---

Fixes #665

sindresorhus commented on source/npm/util.js:
```diff
@@ -68,13 +68,13 @@ exports.collaborators = async pkg => {
 
 	try {
-		const {stdout} = await execa('npm', args);
+		const {stdout} = await execa('npm access list collaborators', args);
```
This is not a correct use of `execa`.

sindresorhus commented:
> const args = ['access', 'list collaborators', packageName];

You have a typo. Should be:

```js
const args = ['access', 'list', 'collaborators', packageName];
```

sindresorhus commented:
> **Make sure that npm test passes locally** and to review your own diff

DerTimonius commented:
Thank you for your feedback!
I had troubles running the tests on my (windows) machine yesterday, found a workaround but that clearly did not work.
I changed the line as you suggested, I also changed two tests of the ``prerequisites-tasks.js`` because it used the old ``npm access`` subcommands (not committed yet).
But running the tests (with the help of ``cross-env``) now, I'm getting the following error with this ``index » version is invalid`` and ``index » version is pre-release``:
```bash
TimeoutError {
     context: {},
     message: 'Connection to npm registry timed out',
 }
```

DerTimonius commented:
So, I now committed the changes to the test files as well and now the Github Actions tests run without error on my forked repo 👍

ar1a commented on source/npm/util.js:
```diff
@@ -68,7 +68,7 @@ exports.collaborators = async pkg => {
 	ow(packageName, ow.string);
 
-	const args = ['access', 'ls-collaborators', packageName];
+	const args = ['access', 'list', 'collaborators', packageName];
```
this needs a `--json` as well

schuchard commented:
Hey @DerTimonius, looks like you're pretty close. Anything I can do to help?

DerTimonius commented:
Hey @schuchard , thank you for your offer!
I did not have time in the last two days to take a look into this again, and now I added the --json flag. I'm not sure if I missed anything else though, so if I did, any help would be appreciated 😃

schuchard commented:
Hi @sindresorhus, is there anything else needed to merge this PR? Thanks.

DerTimonius commented:
I don't know why the tests are failing now with an error message that should be solved (`npm access list not a valid argument`).

karlhorky commented:
Maybe it's failing on an older version of `npm`? (eg. one that does not yet have `npm access list ...` commands?

schuchard commented:
Seems likely. [NPM 9](https://docs.npmjs.com/cli/v9/commands/npm-access?v=true) brought a [breaking change](https://github.blog/changelog/2022-10-24-npm-v9-0-0-released/) to npm access:
> `npm access` subcommands have been renamed

If this package intends to support npm < 9, conditional logic and tests are needed.

DerTimonius commented:
I'm making progress: I implemented a `npm version` check, so that the correct subcommands are used.

Currently I have one failing test left:
```bash
prerequisite-tasks › should fail when user is not authenticated at external registry

142:   await t.throwsAsync(run(testedModule('2.0.0', {name: 'test', version:…
   143:     {message: 'You do not have write permissions required to publish th…

  Promise resolved with:

  undefined
```
Locally, I always get a Timeout Error on this test though...

DerTimonius commented:
Any hints on what I'm missing?

karlhorky commented:
So this is the error message that you're getting:

```
  1 test failed

  prerequisite-tasks › should fail when user is not authenticated at external registry

  /home/runner/work/np/np/test/prerequisite-tasks.js:145

   144:   process.env.NODE_ENV = 'P';                                           
   145:   await t.throwsAsync(run(testedModule('2.0.0', {name: 'test', version:…
   146:     {message: 'You do not have write permissions required to publish th…

  Promise rejected with unexpected exception:

  Error {
    context: {},
    message: 'Git tag `v2.0.0` already exists.',
  }

  Expected message to equal:

  'You do not have write permissions required to publish this package.'

  Object.exports.verifyTagDoesNotExistOnRemote (source/git-util.js:206:9)
  Task.task (source/prerequisite-tasks.js:88:5)
```

This indicates that the collaborators JSON here actually passed the test:

https://github.com/sindresorhus/np/blob/95622c0adbd64920b3d86576784460dd41b81d0e/source/prerequisite-tasks.js#L44-L48

You could try logging out this value to see if this is truly the case.

Or, second idea: maybe the check for `list collaborators` is not being run at all... 🤔  Seems like you have a weird pattern here:

https://github.com/DerTimonius/np/blob/6d4f03650d7c98ec2c352125eb783d64f2e00f39/test/prerequisite-tasks.js#L11-L14

To me `testArgs` should be an array, not a function...

DerTimonius commented:
Thank you for your inputs @karlhorky !

After logging out a few things, I narrowed the problem down to an issue with the external registry: 
* The collaborators functions return `false` and the task does not get to the permissions variable.
* Logging the `args` inside the collaborators function returns a valid array. 
* But since it returned false, the only place where this could happen is here in `source/npm/util.js`, lines 78-88:
```Javascript
try {
		const {stdout} = await execa('npm', args);
		return stdout;
	} catch (error) {
		// Ignore non-existing package error
		if (error.stderr.includes('code E404')) {
			return false;
		}

		throw error;
	}
```

So, logging out the error here gives me this message:
```
'npm ERR! code E404\n' +
    'npm ERR! 404 Not Found - GET http://my.io/-/package/test/collaborators\n' +
    '\n' +
    'npm ERR! A complete log of this run can be found in:\n'
```
But the command is this:
```
npm access list collaborators --json test --registry http://my.io
```

I don't know where the `-/package/test/collaborators` is coming from, it's not added in the code... (or if this is even the issue).



Regarding this:

> Or, second idea: maybe the check for `list collaborators` is not being run at all... 🤔 Seems like you have a weird pattern here:
> 
> https://github.com/sindresorhus/np/blob/6d4f03650d7c98ec2c352125eb783d64f2e00f39/test/prerequisite-tasks.js#L11-L14
> 
> To me `testArgs` should be an array, not a function...

You're right, I tried to do is as an array in the first place, but wasn't able to for some reason. Put it inside the test now, just to be sure.

DerTimonius commented:
So, I removed the conditional arguments in the tests since it does not have any effect on the executed code, now it's again this

```
command: 'npm access ls-collaborators test',
```

instead of:
```
command: `npm access ${testArgs[0]} test${testArgs[1]}',
```
I suspected that maybe this change could be the reason the tests are failing.
Also I moved the `--json` flag once again.

DerTimonius commented:
Hey @sindresorhus, is there anything else I have to change?
